### PR TITLE
Handle jax.numpy.dtype('bfloat16') properly.

### PR DESCRIPTION
**Pull request by adarob**

### CODE REVIEW

**Comment by jakevdp** [comment]
Looks good, thanks!

**Review by jakevdp — APPROVED** [review]

**Comment by mattjj** [comment]
Can we add a test so we don't accidentally regress this?

**Comment by jakevdp** [comment]
@adarob are you interested in adding a test for this? Thanks!

**Comment by adarob** [comment]
Sure

**Comment by adarob** [comment]
Oddly, the tests passes without this change when I run it locally, but not in Colab. Any ideas about that @mattjj ?

```
assert jax.numpy.dtype('bfloat16') == jax.numpy.bfloat16
```

**Comment by jakevdp** [comment]
That test fails on the master branch. I suspect you somehow installed JAX from your branch somewhere in your Python path, so the tests are hitting the modified code.

**Comment by adarob** [comment]
test added

**Comment by jakevdp** [comment]
Thanks!

File: jax/dtypes.py [hunk 1]
```diff
@@ -325,6 +325,8 @@ def is_python_scalar(x):
     return type(x) in python_scalar_dtypes
 
 def dtype(x):
+  if isinstance(dtype, str) and dtype == "bfloat16":
```
Comment (jakevdp) [inline jax/dtypes.py:328]: I see the issue... this should be `isinstance(x, str) and x == "bfloat16":`

Comment (jakevdp) [reply jax/dtypes.py:328]: After you fix this, could you squash all the changes into a single commit? Thanks!

File: jax/dtypes.py [hunk 2]
```diff
@@ -325,6 +325,8 @@ def is_python_scalar(x):
...
+    return bfloat16
```
Comment (jakevdp) [inline jax/dtypes.py:329]: Also, I just noticed that this should be `np.dtype(bfloat16)` (`result_type` returns a dtype object; `bfloat16` is a type not a dtype)

**Review by hawkinsp — COMMENTED** [review]
FYI: you might also want to know about this PR:
https://github.com/tensorflow/tensorflow/commit/16671ca46eda906e822f61a3979f1f68d8bf603c

which added `np.dtype("bfloat16")` support for classic NumPy. That code is used by both JAX and TF, so it will show up with a jaxlib built from head.

**Comment by jakevdp** [comment]
Shoot, looks like this got merged with failing tests.

I'll send a fix

**Comment by jakevdp** [comment]
Fixed in #6176

**Pull request merged**